wagtail/core/models/__init__.py

```diff
@@ -1119,44 +1135,109 @@ def _get_site_root_paths(self, request=None):
 
-        This is used internally by the ``full_url``, ``url``, ``relative_url``
-        and ``get_site`` properties and methods; pages with custom URL routing
-        should override this method in order to have those operations return
-        the custom URLs.
+        Returns a tuple of root paths for all sites this page belongs to.
+        """
+        return tuple(
+            srp
+            for srp in self._get_site_root_paths(cache_object)
+            if self.url_path.startswith(srp.root_path)
+        )
 
-        Accepts an optional keyword argument ``request``, which may be used
-        to avoid repeated database / cache lookups. Typically, a page model
-        that overrides ``get_url_parts`` should not need to deal with
-        ``request`` directly, and should just pass it to the original method
-        when calling ``super``.
+    def _get_best_site_root_path(self, site=None, request=None):
         """
+        .. versionadded::2.16
 
-        possible_sites = [
-            (pk, path, url, language_code)
-            for pk, path, url, language_code in self._get_site_root_paths(request)
-            if self.url_path.startswith(path)
-        ]
+        Returns the most relevant `SiteRootPath` named tuple for this page,
+        based on the supplied ``site`` and ``request`` values.
+        """
+        cache_object = request
+        if request is None and isinstance(site, Site):
+            cache_object = site
+        options = self._get_relevant_site_root_paths(cache_object)
 
-        if not possible_sites:
+        if site:
+            # find an exact match for `site`, or error
+            if isinstance(site, int):
+                site_pk = site
+            else:
+                site_pk = site.pk
+            for option in options:
+                if option.site_id == site_pk:
+                    return option
+            raise ValueError(f"Page '{self}' does not belong to site '{site}'.")
+
+        if not options:
+            # the page does not belong to any sites
             return None
 
-        site_id, root_path, root_url, language_code = possible_sites[0]
+        if len(options) > 1 and isinstance(request, HttpRequest):
+            # attempt to return a more relevant option by deriving ``Site``
+            # from the request, but do not error if unsuccessful
+            site = Site.find_for_request(request)
+            if site:
+                for option in options:
+                    if option.site_id == site.pk:
+                        return option
 
-        site = Site.find_for_request(request)
-        if site:
-            for site_id, root_path, root_url, language_code in possible_sites:
-                if site_id == site.pk:
-                    break
-            else:
-                site_id, root_path, root_url, language_code = possible_sites[0]
+        # fallback to the first item
+        return options[0]
 
+    def get_url_parts(self, request=None, site=None):
```

Comment:
This refactoring of core routing logic is quite extensive. I think it would be best to do this in a separate PR so it could be more carefully considered

Reply:
With all due respect, it's been a chore getting this far, and I am realistically out of time. I would very much appreciate it if we could do the considering here, with the changes all together?